shards/shards.go

```diff
@@ -535,9 +543,34 @@ func (ss *shardedSearcher) streamSearch(ctx context.Context, proc *process, q qu
 					metricSearchMatchCountTotal.Add(float64(sr.Stats.MatchCount))
 					metricSearchNgramMatchesTotal.Add(float64(sr.Stats.NgramMatches))
 
+					// MaxPendingShardPriority *cannot* be this result's ShardPriority, because
```

Comment:
It would be good to describe at a high level how this works in addition to these invariants and potential race conditions. For instance, it took me a bit to realise that correctness relies on shards being already sorted by descending priority.